streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -17,242 +17,225 @@
-                    // else if topic exists with more #.partitions than needed, delete in order to re-create it
-                    try {
-                        deleteTopic(topic);
-                    } catch (ZkNodeExistsException e) {
-                        // ignore and continue
-                    }
-                } else if (topicMetadata.size() < numPartitions) {
-                    // else if topic exists with less #.partitions than needed, add partitions
-                    try {
-                        addPartitions(topic, numPartitions - topicMetadata.size(), replicationFactor, topicMetadata);
-                    } catch (ZkNoNodeException e) {
-                        // ignore and continue
-                    }
-                } else {
-                    topicNotReady = false;
-                }
-            }
-        }
+    public InternalTopicManager(StreamsConfig config, int replicationFactor) {
+        this.config = config;
+        this.streamsKafkaClient = new DefaultKafkaClientSupplier().getStreamKafkaClient(config);
+        this.kafkaClient = streamsKafkaClient.getKafkaClient();
+        this.replicationFactor = replicationFactor;
     }
 
-    private List<Integer> getBrokers() {
-        List<Integer> brokers = new ArrayList<>();
-        for (String broker: zkClient.getChildren(ZK_BROKER_PATH)) {
-            brokers.add(Integer.parseInt(broker));
-        }
-        Collections.sort(brokers);
-
-        log.debug("Read brokers {} from ZK in partition assignor.", brokers);
 
-        return brokers;
+    public void shutDown() throws IOException {
+        this.streamsKafkaClient.shutdown();
     }
+    public void makeReady(String topic, int numPartitions, boolean compactTopic) {
 
-    @SuppressWarnings("unchecked")
-    private Map<Integer, List<Integer>> getTopicMetadata(String topic) {
-        String data = zkClient.readData(ZK_TOPIC_PATH + "/" + topic, true);
-
-        if (data == null) return null;
+        Node brokerNode = kafkaClient.leastLoadedNode(new SystemTime().milliseconds());
 
-        try {
-            ObjectMapper mapper = new ObjectMapper();
+        // Delete the topic if it exists before.
+        boolean topicExists = topicExists(topic, brokerNode);
+        if (topicExists) {
+            deleteTopic(topic, brokerNode);
+            log.debug("Deleted: " + topic);
+        }
+        // Create a new topic
+        createTopic(topic, numPartitions, replicationFactor, compactTopic, brokerNode);
+        log.debug("Created: " + topic);
+    }
 
-            Map<String, Object> dataMap = mapper.readValue(data, new TypeReference<Map<String, Object>>() {
 
-            });
+    /**
+     * Check if the topic exists already
+     *
+     * @param topic
+     * @param brokerNode
+     * @return
+     */
+    private boolean topicExists(String topic, Node brokerNode) {
+        MetadataRequest metadataRequest = new MetadataRequest(Arrays.asList(topic));
```

Comment:
I'm wondering if a lot of this code should be in `StreamsKafkaClient` or perhaps remove `StreamsKafkaClient` completely and construct the underlying `NetworkClient` in `InternalTopicManager`. If the functionality is something we can/will use elsewhere then I'd put it in `StreamsKafkaClient`